adal/src/main/java/com/microsoft/aad/adal/Discovery.java

```diff
@@ -101,18 +102,29 @@ void validateAuthorityADFS(final URL authorizationEndpoint, final String domain)
             throw new IllegalArgumentException("Cannot validate AD FS Authority with domain [null]");
         }
         validateADFS(authorizationEndpoint, domain);
-        validateAuthority(authorizationEndpoint);
     }
 
     void validateAuthority(final URL authorizationEndpoint) throws AuthenticationException {
         verifyAuthorityValidInstance(authorizationEndpoint);
 
-        if (!VALID_HOSTS.contains(authorizationEndpoint.getHost().toLowerCase(Locale.US))) {
-            // host can be the instance or inside the validated list.
-            // Valid hosts will help to skip validation if validated before
-            // call Callback and skip the look up
-            // Only query from Prod instance for now, not all of the instances in the list
-            queryInstance(authorizationEndpoint);
+        if (AuthorityValidationMetadataCache.containsAuthorityHost(authorizationEndpoint)) {
+            return;
+        }
+
+        final String authorityHost = authorizationEndpoint.getHost().toLowerCase(Locale.US);
+        final String trustedHost;
+        if (AAD_WHITELIST_HOSTS.contains(authorizationEndpoint.getHost().toLowerCase(Locale.US))) {
+            trustedHost = authorityHost;
+        } else {
+            trustedHost = TRUSTED_QUERY_INSTANCE;
+        }
+
+        try {
+            sLock = getLock();
+            sLock.lock();
+            performInstanceDiscovery(authorizationEndpoint, trustedHost);
```

Comment:
this is looking dangerous. You took a lock and then made a network request, say the request blocks up for 10 minutes (like we have seen previously). It will block every other validateAuthority request. What is the downside of making multiple discovery requests?

Reply:
Thanks for bringing this up. I was actually planning to talk to Ryan about adding timeout.

Reply:
One network request at a time for authority validation is by design. The issue we have seen before is about write timeout, for which we cannot set a default value. Instance discovery is a Get request, we should be fine for this case.

Reply:
Conceptually, the lock should be protecting solely the data structure, and not the entire authority validation request. A serialized task queue should be used to ensure that only a singular authority validation request is made at any given moment.

What are the side effects of calling a blocking operation on the main thread on Android? The documentation for this method says:

```
If the lock is held by another thread then the current thread becomes disabled for thread scheduling purposes and lies dormant until the lock has been acquired, at which time the lock hold count is set to one.
```

Which my interpretation of would mean if you called it on the main UI thread it would sleep the thread until the lock was released... which would probably not be what you want.

Reply:
@RPangrle calling network on UI thread will crash it. My concern was if the lock was taken and we hit the network and that call gets delayed or stuck, all of that apps threads doing auth will get stuck behind this one lock. Which will not be a deadlock, but still a slowdown.
@weijjia I understand that you mentioned that GET has a timeout, but Ryan mentions here that the intention of the lock was just to protect the data structure. Can we change the code to reflect that?

Reply:
@addev-ashish so first there is no network request made on main thread.  Second, what would be delayed? read timeout and connection timeout is 30 seconds. Also, data structure is already protected by concurrent map, how we are going to make sure one request goes out at a time? The lock is solely for making sure network call for only instance discovery request. You have to provide me concrete examples showing your concern. Only worry won't help here. If it helps for us to go back to the design and figure out what the design is, I'm happy to do that.